In [224]:
import pandas as pd

### Задание

#### Описание полей
##### surgery?
      1 = Yes, it had surgery
      2 = It was treated without surgery
##### Age
      1 = Adult horse
      9 = Young (< 6 months)
##### rectal temperature
      - linear
      - in degrees celsius.
      - An elevated temp may occur due to infection.
      - temperature may be reduced when the animal is in late shock
      - normal temp is 37.8
      - this parameter will usually change as the problem progresses
           eg. may start out normal, then become elevated because of the lesion, passing back through the normal range as
           the horse goes into shock   
##### pulse
      - linear
      - the heart rate in beats per minute
      - is a reflection of the heart condition: 30 -40 is normal for adults
      - rare to have a lower than normal rate although athletic horses may have a rate of 20-25
      - animals with painful lesions or suffering from circulatory shock may have an elevated heart rate
##### respiratory rate
      - linear
      - normal rate is 8 to 10
      - usefulness is doubtful due to the great fluctuations
##### temperature of extremities
      - a subjective indication of peripheral circulation
      - possible values:
           1 = Normal
           2 = Warm
           3 = Cool
           4 = Cold
      - cool to cold extremities indicate possible shock
      - hot extremities should correlate with an elevated rectal temp.
##### pain 
      - a subjective judgement of the horse's pain level
      - possible values:
           1 = alert, no pain
           2 = depressed
           3 = intermittent mild pain
           4 = intermittent severe pain
           5 = continuous severe pain
      - should NOT be treated as a ordered or discrete variable!
      - In general, the more painful, the more likely it is to require surgery
      - prior treatment of pain may mask the pain level to some extent
##### outcome
      - what eventually happened to the horse?
      - possible values:
           1 = lived
           2 = died
           3 = was euthanized

Поля surgery?, Age, temperature of extremities, pain и outcome содержат *качественные данные*.

Поля rectal temperature, pulse и respiratory rate содержат *количественные непрерывные величины*.


In [225]:
# получение датафрейма
horse_data = pd.read_csv('horse_data.csv', names = ['surgery?', 'Age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces', 'abdomen', 'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome', 'surgical lesion?', 'site of lesion', 'type of lesion', 'subtype of lesion', 'specific code of lesion', 'cp_data'], na_values = '?')
horse_data.head(10)

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0       2.0    1           530101                38.5   66.0   
1       1.0    1           534817                39.2   88.0   
2       2.0    1           530334                38.3   40.0   
3       1.0    9          5290409                39.1  164.0   
4       2.0    1           530255                37.3  104.0   
5       2.0    1           528355                 NaN    NaN   
6       1.0    1           526802                37.9   48.0   
7       1.0    1           529607                 NaN   60.0   
8       2.0    1           530051                 NaN   80.0   
9       2.0    9          5299629                38.3   90.0   

   respiratory rate  temperature of extremities  peripheral pulse  \
0              28.0                         3.0               3.0   
1              20.0                         NaN               NaN   
2              24.0                         1.0               1.0   
3              84.0                         4.0               1.0   
4              35.0                         NaN               NaN   
5               NaN                         2.0               1.0   
6              16.0                         1.0               1.0   
7               NaN                         3.0               NaN   
8              36.0                         3.0               4.0   
9               NaN                         1.0               NaN   

   mucous membranes  capillary refill time  ...  total protein  \
0               NaN                    2.0  ...            8.4   
1               4.0                    1.0  ...           85.0   
2               3.0                    1.0  ...            6.7   
3               6.0                    2.0  ...            7.2   
4               6.0                    2.0  ...            7.4   
5               3.0                    1.0  ...            NaN   
6               1.0                    1.0  ...            7.0   
7               NaN                    1.0  ...            8.3   
8               3.0                    1.0  ...            6.2   
9               1.0                    1.0  ...            6.2   

   abdominocentesis appearance  abdomcentesis total protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   
5                          NaN                          NaN      1.0   
6                          NaN                          NaN      1.0   
7                          NaN                          NaN      2.0   
8                          NaN                          NaN      3.0   
9                          1.0                          2.2      1.0   

   surgical lesion?  site of lesion  type of lesion  subtype of lesion  \
0                 2           11300               0                  0   
1                 2            2208               0                  0   
2                 2               0               0                  0   
3                 1            2208               0                  0   
4                 2            4300               0                  0   
5                 2               0               0                  0   
6                 1            3124               0                  0   
7                 1            2208               0                  0   
8                 1            3205               0                  0   
9                 2               0               0                  0   

   specific code of lesion  cp_data  
0                        2      NaN  
1                        2      NaN  
2                        1      NaN  
3                        1      NaN  
4            

In [226]:
# фильтрация обозначенных 8 столбцов
filtered_horse_data = horse_data[['surgery?', 'Age','rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']]
filtered_horse_data.head(10)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   
5       2.0    1                 NaN    NaN               NaN   
6       1.0    1                37.9   48.0              16.0   
7       1.0    1                 NaN   60.0               NaN   
8       2.0    1                 NaN   80.0              36.0   
9       2.0    9                38.3   90.0               NaN   

   temperature of extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0  
5                         2.0   2.0      1.0  
6                         1.0   3.0      1.0  
7                         3.0   NaN      2.0  
8                         3.0   4.0      3.0  
9                         1.0   5.0      1.0

In [227]:
# описательная статистика
filtered_horse_data.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [228]:
print("Мода поля surgery?:", filtered_horse_data['surgery?'].mode()[0])
print("Мода поля Age:", filtered_horse_data['Age'].mode()[0])
print("Мода поля rectal temperature:", filtered_horse_data['rectal temperature'].mode()[0])
print("Мода поля pulse:", filtered_horse_data['pulse'].mode()[0])
print("Мода поля respiratory rate:", filtered_horse_data['respiratory rate'].mode()[0])
print("Мода поля temperature of extremities:", filtered_horse_data['temperature of extremities'].mode()[0])
print("Мода поля pain:", filtered_horse_data['pain'].mode()[0])
print("Мода поля outcome:", filtered_horse_data['outcome'].mode()[0])

Мода поля surgery?: 1.0
Мода поля Age: 1
Мода поля rectal temperature: 38.0
Мода поля pulse: 48.0
Мода поля respiratory rate: 20.0
Мода поля temperature of extremities: 3.0
Мода поля pain: 3.0
Мода поля outcome: 1.0


#### Базовая описательная статистика
##### Мода по полям
чаще всего:
1) лошадей оперировали, чем лечили без хирургического вмещательства,
2) были взрослые особи,
3) наблюдалась чуть повышенная ректальная температура (38.0),
4) наблюдался чуть повышенный пульс (48.0),
5) наблюдалась повышенная частота дыхания (20.0),
6) наблюдалась прохладная температура конечностей,
7) наблюдалась периодическая легкая боль,
8) лошади выживали.

##### Медиана по полям
Медианные значения:
1) повышенная ректальная температура (38.2),
2) учащенный пульс (64.0),
3) учащенное дыхание (24.5).

##### Среднее по полям
В среднем наблюдались:
1) повышенная ректальная температура (38.17),
2) учащенный пульс (71.9),
3) учащенное дыхание (30.42).

##### Стандартное отклонение
В среднем разброс величин от средней:
1) 0.73 градуса по ректальной температуре (от 38.17),
2) 28.63 удара/мин по пульсу (от 71.9),
3) 17.64 вдоха/мин по частоте дыхания (от 30.42).



In [229]:
## выбросы
# ректальная температура
q1 = filtered_horse_data['rectal temperature'].quantile(0.25)
q3 = filtered_horse_data['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers_rec_temp = filtered_horse_data[filtered_horse_data['rectal temperature'].between(lower_bound, upper_bound, inclusive = True)]
concatenated_data_rec_temp = pd.concat([filtered_horse_data,remove_outliers_rec_temp])
concatenated_data_rec_temp[~pd.isna(concatenated_data_rec_temp['rectal temperature'])].drop_duplicates(keep = False).sort_values(by = 'rectal temperature')

surgery?  Age  rectal temperature  pulse  respiratory rate  \
44        1.0    1                35.4  140.0              24.0   
141       2.0    1                36.0   42.0              30.0   
238       2.0    1                36.1   88.0               NaN   
80        1.0    1                36.4   98.0              35.0   
118       1.0    1                36.5   78.0              30.0   
298       1.0    1                36.5  100.0              24.0   
251       2.0    1                36.6   42.0              18.0   
99        2.0    1                39.6  108.0              51.0   
75        1.0    9                39.7  100.0               NaN   
20        1.0    1                39.9   72.0              60.0   
281       2.0    1                40.0   78.0               NaN   
54        2.0    1                40.3  114.0              36.0   
91        2.0    1                40.3  114.0              36.0   
259       1.0    1                40.8   72.0              42.0   

     temperature of extremities  pain  outcome  
44                          3.0   4.0      3.0  
141                         NaN   NaN      2.0  
238                         3.0   3.0      3.0  
80                          3.0   4.0      2.0  
118                         1.0   5.0      1.0  
298                         3.0   3.0      1.0  
251                         3.0   1.0      2.0  
99                          3.0   2.0      1.0  
75                          3.0   2.0      3.0  
20                          1.0   5.0      1.0  
281                         3.0   2.0      2.0  
54                          3.0   2.0      3.0  
91                          3.0   2.0      2.0  
259                         3.0   2.0      2.0

In [230]:
# пульс
q1 = filtered_horse_data['pulse'].quantile(0.25)
q3 = filtered_horse_data['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers_pulse = filtered_horse_data[filtered_horse_data['pulse'].between(lower_bound, upper_bound, inclusive = True)]
concatenated_data_pulse = pd.concat([filtered_horse_data,remove_outliers_pulse])
concatenated_data_pulse[~pd.isna(concatenated_data_pulse['pulse'])].drop_duplicates(keep = False).sort_values(by = 'pulse')

surgery?  Age  rectal temperature  pulse  respiratory rate  \
41        2.0    9                39.0  150.0              72.0   
275       1.0    9                38.8  150.0              50.0   
55        1.0    9                38.6  160.0              20.0   
3         1.0    9                39.1  164.0              84.0   
255       1.0    9                38.8  184.0              84.0   

     temperature of extremities  pain  outcome  
41                          NaN   NaN      1.0  
275                         1.0   5.0      2.0  
55                          3.0   3.0      2.0  
3                           4.0   2.0      2.0  
255                         1.0   4.0      2.0

In [231]:
# частота дыхания
q1 = filtered_horse_data['respiratory rate'].quantile(0.25)
q3 = filtered_horse_data['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers_resp_rate = filtered_horse_data[filtered_horse_data['respiratory rate'].between(lower_bound, upper_bound, inclusive = True)]
concatenated_data_resp_rate = pd.concat([filtered_horse_data,remove_outliers_resp_rate])
concatenated_data_resp_rate[~pd.isna(concatenated_data_resp_rate['respiratory rate'])].drop_duplicates(keep = False).sort_values(by = 'respiratory rate')

surgery?  Age  rectal temperature  pulse  respiratory rate  \
120       1.0    1                39.4   54.0              66.0   
103       1.0    9                38.0  140.0              68.0   
125       1.0    1                38.0   42.0              68.0   
265       2.0    1                 NaN   56.0              68.0   
229       1.0    9                38.5  120.0              70.0   
295       1.0    1                 NaN  120.0              70.0   
41        2.0    9                39.0  150.0              72.0   
82        1.0    9                38.1  100.0              80.0   
84        1.0    1                37.8   60.0              80.0   
208       1.0    1                37.8   88.0              80.0   
3         1.0    9                39.1  164.0              84.0   
255       1.0    9                38.8  184.0              84.0   
244       1.0    9                38.2  124.0              88.0   
39        1.0    9                39.2  146.0              96.0   
106       1.0    1                38.3   52.0              96.0   

     temperature of extremities  pain  outcome  
120                         1.0   2.0      1.0  
103                         1.0   3.0      1.0  
125                         4.0   3.0      1.0  
265                         3.0   3.0      3.0  
229                         NaN   NaN      1.0  
295                         4.0   2.0      3.0  
41                          NaN   NaN      1.0  
82                          3.0   3.0      1.0  
84                          1.0   2.0      1.0  
208                         3.0   NaN      3.0  
3                           4.0   2.0      2.0  
255                         1.0   4.0      2.0  
244                         1.0   2.0      1.0  
39                          NaN   NaN      2.0  
106                         NaN   NaN      1.0

#### Выбросы
При анализе выбросов отфильтрованы пустые значения.

##### Ректальная температура
Согласно описанию, норма = 37,8. 
Повышение температуры сигнализирует о поражении органов.
Понижение температуры сигнализирует о позднем шоке. 

Экстремально низкие температуры сопряжены с постоянной сильной болью, значительно повышенном пульсом и с последующей эвнатаназией животного или его естественной смертью. Поэтому данные не могут быть удалены как ошибочные выбросы и являются естественными данными, соответствующими сделанным предположениям.

Экстремально высокие температуры сопряжены с значительно повышенном пульсом и с последующей эвнатаназией животного или его естественной смертью. Поэтому данные не могут быть удалены как ошибочные выбросы и являются естественными данными, соответствующими сделанным предположениям.

##### Пульс
Согласно описанию, норма = 30-40 ударов / мин. 
Более низкий уровень характерен для атлетичных лошадей (в данном датасете таких выбросов нет).
Повышенный уровень пульса может говорить о поражении органов или шоке животного. 

Экстремально высокие температуры наблюдаются только у молодых особей (до 6 месяцев), у каждой особи отмечена повышенная температура и болевой синдром различной тяжести. В 4 из 5 случаев не было хирургического вмешательства и особь скончалась. Несмотря на значительно превышающие средние значения показатели, данные не должны быть удалены как ошибочные выбросы, поскольку могут быть аномальными, но естественными данными, необходимыми для корректного анализа.

##### Частота дыхания
Согласно описанию, норма = 8-10. 
Использование как значимого фактора сомнительно в силу больших флуктуаций данных.

Анализ выбросов также сложен из-за большого разброса данных. Более того, данное поле одно из наименее заполненных в датасете. Поэтому как пропуски, так и выбросы здесь не являются значимыми в анализе датасета.

##### Прочие характеристики
Остальные данные являются качественными показателями и принимают значения в пределах обозначенных значений. Проверка на выбросы бессмысленны.

In [279]:
# пропущенные значения - абсолютные величины
filtered_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [290]:
# пропущенные значения - в %
(filtered_horse_data.isna().mean()*100).round(2)

surgery?                       0.33
Age                            0.00
rectal temperature            20.00
pulse                          8.00
respiratory rate              19.33
temperature of extremities    18.67
pain                          18.33
outcome                        0.33
dtype: float64

Поля surgery? и outcome почти полностью заполнены и имеют только по 1 пропуску (0.33% всех значений).
Наибольшее число пропусков - в данных по ректальной температуре (20%), частоте дыхания (19.33%), температуре конечностей (18.67%) и боли (18.33%).
Поле Age полностью заполнено.
При удалении всех пропусков происходит потеря половины данных, что недопустимо.

#### Обработка пропусков
##### surgery? и outcome
В полях пропущено по одному значению и эти значения соответствуют одному наблюдению. Замена этих пропусков расчетными или аналитическими значениями невозможна, поскольку это исказит данные. Удаление одной строки из 300 в данной ситуации исказит данные незначительно.

##### частота дыхания
Поскольку поле не является значимым для анализа из-за сильной флуктуации, данное поле может быть полностью удалено при наличии в нем пропусков.

##### температура, пульс, температура конечностей, боль
В датасете присутствуют строки, где не заполнена информация по выше перечисленным полям. Данные пропуски не могут быть заполнены средними значениями, поскольку в разрезе данных о хирургическом вмешательстве, возрасте и исходе значения разнятся. Удаление этих строк приведет к потере всего 3% данных, что в принятых пределах достоверности в 95% можно считать допустимым.

##### пульс
По описанию, при болезненном поражении органов и острой сердечной недостаточности может наблюдаться повышенный уровень пульса. Поэтому для более точного определения пропущенных значений пульса взята группировка по боли. 
Также было замечено, что пульс разнится для лошадей разных возрастных групп и в зависимости от хирургического вмешательства, поэтому данные показатели также были учетны.
Пустоты заполнены медианными значениями по группировке: возраст, хирургическое вмешательство, боль.

##### температура
По описанию, повышенная температура наблюдается вследствие инфекции, а пониженная вследствие сердечной недостаточности. Однако отклонения замечены для различных возрастов и исходов для животного.
Пустоты заполнены медианными значениями по группировке: возраст, исход.

##### температура конечностей
По описанию, горячие конечности должны коррелировать с повышенной температурой. На данных это однозначно не подтверждается.
Пустоты заполнены медианными значениями по группировке: исход, температура.
Таким образом заполнилось 31 пустое значение из 48.
Для дозаполнения применено медианное значение по группировке: возраст, исход, боль, хирургическое вмешательство.

##### боль
По описанию, в обще случае, чем выше боль, тем более вероятно, что понадобится хирургическое вмешательство, что частично доказывается данными. Также выявлена некая взаимосвязь с исходом животного и его возрастом.
Заполнение столь субъективных полей расчетными, однако, остается под сомнением.
Пустоты заполнены медианными значениями по группировке: возраст, хирургическое вмешательство, исход.

В результате заполнения пропусков отмечены незначительные изменения в значениях средней, моды и медианы.

In [291]:
# удаление столбца с частотой дыхания
dealing_with_na_horse_data = filtered_horse_data.drop(['respiratory rate'], axis = 'columns')
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   temperature of extremities  244 non-null    float64
 5   pain                        245 non-null    float64
 6   outcome                     299 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 16.5 KB


In [292]:
# удаление пустых строк по данным о хирургическом вмешательстве и исходе
dealing_with_na_horse_data = dealing_with_na_horse_data.dropna(subset = ['surgery?','outcome'])
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 299 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         299 non-null    int64  
 2   rectal temperature          239 non-null    float64
 3   pulse                       275 non-null    float64
 4   temperature of extremities  243 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     299 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.7 KB


In [293]:
# удаление строк, где пусты значения температуры, пульса, температуры конечностей и боли
dealing_with_na_horse_data = dealing_with_na_horse_data.dropna(thresh = 1, subset = ['rectal temperature', 'pulse','temperature of extremities','pain'])
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          239 non-null    float64
 3   pulse                       275 non-null    float64
 4   temperature of extremities  243 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [237]:
# данные для rectal temperature
filtered_horse_data.groupby(['Age','surgery?','pain']).agg('median')

rectal temperature  pulse  respiratory rate  \
Age surgery? pain                                                
1   1.0      1.0                38.10   48.0              16.0   
             2.0                38.30   80.0              36.0   
             3.0                38.05   60.0              22.0   
             4.0                38.10   72.0              30.0   
             5.0                38.00   80.0              30.0   
    2.0      1.0                38.30   48.0              20.0   
             2.0                38.30   60.0              20.0   
             3.0                38.20   54.0              25.0   
             4.0                37.90   83.0              24.5   
             5.0                38.50   96.0              28.0   
9   1.0      2.0                38.80  124.0              72.0   
             3.0                38.40  120.0              40.0   
             4.0                38.80  156.0              60.0   
             5.0                38.45  143.0              49.0   
    2.0      1.0                38.00   92.0              28.0   
             4.0                  NaN  100.0              44.0   
             5.0                38.30   90.0               NaN   

                   temperature of extremities  outcome  
Age surgery? pain                                       
1   1.0      1.0                          1.0      1.0  
             2.0                          3.0      2.0  
             3.0                          2.0      1.0  
             4.0                          3.0      2.0  
             5.0                          3.0      2.0  
    2.0      1.0                          2.0      1.0  
             2.0                          3.0      1.0  
             3.0                          2.0      1.0  
             4.0                          3.0      3.0  
             5.0                          3.0      2.0  
9   1.0      2.0                          3.0      2.0  
             3.0                          3.0      1.0  
             4.0                          2.0      2.0  
             5.0                          2.0      2.0  
    2.0      1.0                          1.0      2.0  
             4.0                          2.0      1.0  
             5.0                          1.0      1.0

In [238]:
# данные для pulse
filtered_horse_data.groupby(['Age','outcome']).agg('median')

surgery?  rectal temperature  pulse  respiratory rate  \
Age outcome                                                          
1   1.0           1.0                38.2   52.0              24.0   
    2.0           1.0                38.0   82.0              28.0   
    3.0           1.0                38.0   81.0              24.0   
9   1.0           1.0                38.4  120.0              64.0   
    2.0           1.0                38.8  141.0              44.0   
    3.0           1.0                39.7  100.0               NaN   

             temperature of extremities  pain  
Age outcome                                    
1   1.0                             2.0   3.0  
    2.0                             3.0   4.0  
    3.0                             3.0   3.5  
9   1.0                             1.0   3.0  
    2.0                             3.0   3.0  
    3.0                             3.0   2.0

In [265]:
# данные для temperature of extremities

filtered_horse_data.groupby(['Age','outcome','pain','surgery?']).agg('median')

rectal temperature  pulse  respiratory rate  \
Age outcome pain surgery?                                                
1   1.0     1.0  1.0                    38.10   48.0              16.0   
                 2.0                    38.30   48.0              20.0   
            2.0  1.0                    38.30   64.0              36.0   
                 2.0                    38.20   53.0              18.0   
            3.0  1.0                    38.15   56.0              21.0   
                 2.0                    38.20   52.0              26.0   
            4.0  1.0                    38.25   88.0              32.0   
                 2.0                    38.00   40.0              25.0   
            5.0  1.0                    37.90   72.0              30.0   
                 2.0                      NaN  112.0              24.0   
    2.0     1.0  2.0                    36.60   42.0              18.0   
            2.0  1.0                    38.90   80.0              22.0   
                 2.0                    40.00   96.0              36.0   
            3.0  1.0                    37.50   84.0              36.0   
                 2.0                    38.00   86.0              16.0   
            4.0  1.0                    38.10   67.0              24.0   
                 2.0                    37.20   84.0              48.0   
            5.0  1.0                    38.00   89.0              29.0   
                 2.0                    38.50   96.0              28.0   
    3.0     1.0  2.0                    37.60   48.0              20.0   
            2.0  1.0                    37.80   84.0              38.0   
                 2.0                    38.35  111.0              18.0   
            3.0  1.0                    39.20   88.0              20.0   
                 2.0                    37.00   60.0              44.0   
            4.0  1.0                    38.15   80.0              28.0   
                 2.0                    38.15   91.0              24.0   
            5.0  1.0                    38.05   68.0              23.0   
9   1.0     2.0  1.0                    38.25  127.0              74.0   
            3.0  1.0                    38.10  120.0              68.0   
            4.0  2.0                      NaN  100.0              44.0   
            5.0  2.0                    38.30   90.0               NaN   
    2.0     1.0  2.0                    38.00   92.0              28.0   
            2.0  1.0                    38.95  144.0              60.0   
            3.0  1.0                    38.50  122.0              30.0   
            4.0  1.0                    38.80  156.0              60.0   
            5.0  1.0                    38.45  143.0              49.0   
    3.0     2.0  1.0                    39.70  100.0               NaN   

                           temperature of extremities  
Age outcome pain surgery?                              
1   1.0     1.0  1.0                              1.0  
                 2.0                              2.0  
            2.0  1.0                              2.0  
                 2.0                              2.0  
            3.0  1.0                              2.0  
                 2.0                              1.5  
            4.0  1.0                              3.0  
                 2.0                              NaN  
            5.0  1.0                              3.0  
                 2.0                              3.0  
    2.0     1.0  2.0                              3.0  
            2.0  1.0                              3.0  
                 2.0                              3.0  
            3.0  1.0                              2.0  
                 2.0                              3.0  
            4.0  1.0                              1.0  
                 2.0                              3.0  
            5.0  1.0                              3.0  
                 2.0

In [240]:
# данные для pain
filtered_horse_data.groupby(['Age','outcome','surgery?']).agg('median')

rectal temperature  pulse  respiratory rate  \
Age outcome surgery?                                                
1   1.0     1.0                    38.20   60.0              24.0   
            2.0                    38.20   50.0              24.0   
    2.0     1.0                    38.00   73.5              26.0   
            2.0                    38.00   90.0              30.0   
    3.0     1.0                    38.00   79.0              32.0   
            2.0                    38.20   85.0              20.0   
9   1.0     1.0                    38.25  122.0              69.0   
            2.0                    38.80   90.0              37.0   
    2.0     1.0                    38.80  146.0              48.0   
            2.0                    38.00   92.0              28.0   
    3.0     1.0                    39.70  100.0               NaN   

                      temperature of extremities  pain  
Age outcome surgery?                                    
1   1.0     1.0                              3.0   3.0  
            2.0                              2.0   2.0  
    2.0     1.0                              3.0   4.0  
            2.0                              3.0   4.0  
    3.0     1.0                              3.0   4.0  
            2.0                              3.0   3.0  
9   1.0     1.0                              1.0   3.0  
            2.0                              1.5   4.5  
    2.0     1.0                              3.0   3.5  
            2.0                              1.0   1.0  
    3.0     1.0                              3.0   2.0

In [294]:
# заполнение пустоты по пульсу
dealing_with_na_horse_data['pulse'].fillna(filtered_horse_data.groupby(['Age','surgery?','pain'])['pulse'].transform('median'), inplace = True)
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          239 non-null    float64
 3   pulse                       291 non-null    float64
 4   temperature of extremities  243 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [295]:
# заполнение пустоты по температуре
dealing_with_na_horse_data['rectal temperature'].fillna(filtered_horse_data.groupby(['Age','outcome'])['rectal temperature'].transform('median'), inplace = True)
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          291 non-null    float64
 3   pulse                       291 non-null    float64
 4   temperature of extremities  243 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [296]:
# заполнение пустоты по температуре конечностей
dealing_with_na_horse_data['temperature of extremities'].fillna(filtered_horse_data.groupby(['outcome','rectal temperature'])['temperature of extremities'].transform('median').round(0), inplace = True)
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          291 non-null    float64
 3   pulse                       291 non-null    float64
 4   temperature of extremities  274 non-null    float64
 5   pain                        244 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [297]:
# заполнение пустоты по боли
dealing_with_na_horse_data['pain'].fillna(filtered_horse_data.groupby(['Age','surgery?','outcome'])['pain'].transform('median').round(0), inplace = True)
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          291 non-null    float64
 3   pulse                       291 non-null    float64
 4   temperature of extremities  274 non-null    float64
 5   pain                        291 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [300]:
# заполнение пустоты по температуре конечностей
dealing_with_na_horse_data['temperature of extremities'].fillna(dealing_with_na_horse_data.groupby(['Age','outcome','pain','surgery?'])['temperature of extremities'].transform('median').round(0), inplace = True)
dealing_with_na_horse_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 291 entries, 0 to 299
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    291 non-null    float64
 1   Age                         291 non-null    int64  
 2   rectal temperature          291 non-null    float64
 3   pulse                       291 non-null    float64
 4   temperature of extremities  291 non-null    float64
 5   pain                        291 non-null    float64
 6   outcome                     291 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 18.2 KB


In [301]:
dealing_with_na_horse_data.describe()

surgery?         Age  rectal temperature       pulse  \
count  291.000000  291.000000          291.000000  291.000000   
mean     1.405498    1.604811           38.156357   71.996564   
std      0.491834    2.118517            0.668028   28.004864   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.850000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       temperature of extremities        pain     outcome  
count                  291.000000  291.000000  291.000000  
mean                     2.326460    2.986254    1.556701  
std                      1.020457    1.240057    0.742350  
min                      1.000000    1.000000    1.000000  
25%                      1.000000    2.000000    1.000000  
50%                      3.000000    3.000000    1.000000  
75%                      3.000000    4.000000    2.000000  
max                      4.000000    5.000000    3.000000

In [303]:
print("Мода поля surgery?:", dealing_with_na_horse_data['surgery?'].mode()[0])
print("Мода поля Age:", dealing_with_na_horse_data['Age'].mode()[0])
print("Мода поля rectal temperature:", dealing_with_na_horse_data['rectal temperature'].mode()[0])
print("Мода поля pulse:", dealing_with_na_horse_data['pulse'].mode()[0])
print("Мода поля temperature of extremities:", dealing_with_na_horse_data['temperature of extremities'].mode()[0])
print("Мода поля pain:", dealing_with_na_horse_data['pain'].mode()[0])
print("Мода поля outcome:", dealing_with_na_horse_data['outcome'].mode()[0])

Мода поля surgery?: 1.0
Мода поля Age: 1
Мода поля rectal temperature: 38.0
Мода поля pulse: 48.0
Мода поля temperature of extremities: 3.0
Мода поля pain: 3.0
Мода поля outcome: 1.0
